Wallet transaction tab

Jibz1 opened the pull request:
@pchrysochoidis please review.  

- Added Network selector
- Transactions section added. 

RPC endpoints configured in `.env` file
Transaction for a specific address is fetched either on network change or on-page tab switch. Not part of the polling.


https://user-images.githubusercontent.com/8676844/173327388-6c0791db-a325-4b2e-959b-c26e1d13088e.mov

pchrysochoidis commented on wallet/src/ui/app/ApiProvider.ts:
```diff
@@ -54,6 +55,14 @@ export default class ApiProvider {
         this._apiProvider = new JsonRpcProvider(DEFAULT_API_ENDPOINT);
     }
 
+    public setNewJsonRpcProvider(apiEnv: string) {
```
Changing `apiEnv` from `string` to `API_ENV` should make more clear that only `API_ENV` is supported

Jibz1 replied:
👍

pchrysochoidis commented on wallet/src/ui/app/ApiProvider.ts:
```diff
@@ -54,6 +55,14 @@ export default class ApiProvider {
     }
 
+    public setNewJsonRpcProvider(apiEnv: string) {
+        if (apiEnv) {
```
I think if is unnecessary if we use `API_ENV`

Jibz1 replied:
You are right. I take it out.

pchrysochoidis commented on wallet/src/ui/app/components/network-switch/Network.module.scss:
```diff
@@ -0,0 +1,111 @@
+.network {
```
The switch seems a bit too big, especially in the popup view. We could make it a bit smaller less margins etc.

<img width="369" alt="Screenshot 2022-06-13 at 11 22 31" src="https://user-images.githubusercontent.com/10210143/173333499-a74b7cc0-9cd5-4c5e-aed7-482cf5ededf0.png">

Jibz1 replied:
I will make it smaller.  Developed it on the browser version

Jibz1 commented on wallet/src/ui/app/ApiProvider.ts:
```diff
@@ -1,7 +1,8 @@
 // SPDX-License-Identifier: Apache-2.0
 
-import { JsonRpcProvider, RawSigner } from '@mysten/sui.js';
+// import { JsonRpcProvider, RawSigner } from '@mysten/sui.js';
```
I will remove it.

pchrysochoidis commented on wallet/src/ui/app/components/network-switch/NetworkSelector.tsx:
```diff
@@ -0,0 +1,87 @@
+    const selectedApiEnv = useAppSelector(({ app }) => app.apiEnv);
+    const dispatch = useAppDispatch();
+    const networkList: any = API_ENV_TO_INFO;
+    const netWorks = useMemo(
```
We could move that to the constant and avoid doing for every instance of the Network selector. We can move it outside as well but for just those 2 attributes should be fine to go to the `API_ENV_TO_INFO` const

pchrysochoidis commented on wallet/src/ui/app/components/network-switch/NetworkSelector.tsx:
```diff
@@ -0,0 +1,87 @@
+const NetworkSelector = () => {
+    const selectedApiEnv = useAppSelector(({ app }) => app.apiEnv);
+    const dispatch = useAppDispatch();
+    const networkList: any = API_ENV_TO_INFO;
```
I think `any` is not allowed by our linting rules, and in general if possible we should avoid it.

pchrysochoidis commented on wallet/src/ui/app/components/network-switch/NetworkSelector.tsx:
```diff
@@ -0,0 +1,87 @@
+    const changeNetwork = useCallback(
+        (networkName: string) => () => {
+            const apiEnv = API_ENV[networkName as keyof typeof API_ENV];
+            dispatch(setNetworkSelector(true));
```
This feels like it should be a local state to the 'dropdown' component instead of storing it to redux. Maybe we can refactor it to have a generic dropdown component (maybe in another MR)

Jibz1 replied:
Not sure how this will work. Create a local state, listen to the changes in the local state and dispatch an action whenever the is a change?

Jibz1 commented on wallet/src/ui/app/components/network-switch/NetworkSelector.tsx:
```diff
@@ -0,0 +1,87 @@
+            dispatch(setNetworkSelector(true));
+            store.dispatch(setApiEnv(apiEnv));
+            dispatch(changeRPCNetwork()).unwrap();
+            dispatch(getTransactionsByAddress()).unwrap();
```
My bad carried over from the transaction page. Ment for error handling.

pchrysochoidis commented on wallet/src/ui/app/components/network-switch/NetworkSelector.tsx:
```diff
@@ -0,0 +1,87 @@
+        [networkList]
+    );
+
+    const changeNetwork = useCallback(
```
I think we could use only `setApiEnv` and have a middleware to change the provider but also clear the state that is relative with the connected network. Currently there are cases we show coins or nfts that are from another network

Jibz1 replied:
Not sure why we want to show assets from a different network at the same time as the current networks.

pchrysochoidis replied:
I meant that with the current implementation we see objects that are in another network and we shouldn't show them. (eg when we change the network we see coins of the other network until we poll again and hope it is successful!)

pchrysochoidis commented on wallet/src/ui/app/components/network-switch/NetworkSelector.tsx:
```diff
@@ -0,0 +1,87 @@
+        (networkName: string) => () => {
+            const apiEnv = API_ENV[networkName as keyof typeof API_ENV];
+            dispatch(setNetworkSelector(true));
+            store.dispatch(setApiEnv(apiEnv));
```
~~store.~~

pchrysochoidis commented on wallet/src/ui/app/components/network-switch/NetworkSelector.tsx:
```diff
@@ -0,0 +1,87 @@
+                            icon="check2"
+                            className={cl(
+                                st['selected-network'],
+                                selectedApiEnv === apiEnv.networkName
```
Using an object like ``` { [st['network-active']]:  selectedApiEnv === apiEnv.networkName } ``` is a bit more clear I think

pchrysochoidis commented on wallet/src/ui/app/components/network-switch/NetworkSelector.tsx:
```diff
@@ -0,0 +1,87 @@
+                {netWorks.map((apiEnv, index) => (
+                    <li
+                        className={st['network-item']}
+                        key={`networkid-${index}`}
```
key={apiEnv.networkName} is unique no need to use index

pchrysochoidis commented on wallet/src/ui/app/components/network-switch/NetworkSelector.tsx:
```diff
@@ -0,0 +1,87 @@
+                    <li
+                        className={st['network-item']}
+                        key={`networkid-${index}`}
+                        onClick={changeNetwork(apiEnv.networkName)}
```
In general I think is better to use the data fields of the node and avoid creating new functions every time.
```
<li
...
data-network={networkName}
>
```
and access it from the currentTarget of the event

pchrysochoidis commented on wallet/src/ui/app/components/network-switch/NetworkSelector.tsx:
```diff
@@ -0,0 +1,87 @@
+                    </li>
+                ))}
+
+                <li className={st['network-item']}>
```
We should remove/hide this since it's not implemented

pchrysochoidis commented on wallet/src/ui/app/components/network-switch/index.tsx:
```diff
@@ -0,0 +1,77 @@
+    const dispatch = useAppDispatch();
+
+    const openNetworkSelector = useCallback(
+        () => () => {
```
Do we need the extra `() => `?

pchrysochoidis commented on wallet/src/ui/app/components/network-switch/index.tsx:
```diff
@@ -0,0 +1,77 @@
+                <div
+                    className={st.network}
+                    style={netColor}
+                    onClick={openNetworkSelector()}
```
~~()~~

pchrysochoidis commented on wallet/src/ui/app/components/transactions-card/index.tsx:
```diff
@@ -0,0 +1,96 @@
+    kind: TransactionKindName | undefined;
+    From: string;
+};
+const truncate = (txt: string, maxLength: number) => {
```
Why not use the `useMiddleEllipsis` hook?

Jibz1 replied:
`useMiddleEllipsis` Does not work inside the other function

pchrysochoidis commented on wallet/src/ui/app/components/transactions-card/index.tsx:
```diff
@@ -0,0 +1,96 @@
+    )}`;
+};
+
+function TransactionResult({ txresults }: { txresults: TxnData[] }) {
```
Changing this to TransctionListItem or something and move the `map` functionality to a TransactionList where we can show error/loading states would make easier to reuse it maybe, and also use the useMiddleEllipsis hook

pchrysochoidis commented on wallet/src/ui/app/components/transactions-card/index.tsx:
```diff
@@ -0,0 +1,96 @@
+function TransactionResult({ txresults }: { txresults: TxnData[] }) {
+    return (
+        <div className={st['tx-container']}>
+            <h4>
```
I think this should go to the Transactions page

pchrysochoidis commented on wallet/src/ui/app/pages/home/transactions/index.tsx:
```diff
@@ -0,0 +1,24 @@
+
+    return txByAddress && txByAddress.length ? (
+        <TransactionResult
+            txresults={txByAddress.filter((_, index: number) => index <= 4)}
```
`.slice(0, 5)` I think is a bit more clear. And probably doing it at the selector would be better.

Jibz1 replied:
this was done intentionally so that we can have a filter function and filter by other attribute

pchrysochoidis commented on wallet/src/ui/app/redux/RootReducer.ts:
```diff
@@ -7,8 +7,16 @@ import account from './slices/account';
 import suiObjects from './slices/sui-objects';
 import transactions from './slices/transactions';
+// TODO add to transactions slice
+import txresults from './slices/txresults';
```
I don't think we need a new slice, transactions slice is the one we should use to store all the transactions of the user.

Jibz1 replied:
OK

pchrysochoidis commented on wallet/src/ui/app/redux/slices/app/index.ts:
```diff
@@ -1,24 +1,35 @@
+    showHideNetwork: false,
 };
 
+export const changeRPCNetwork = createAsyncThunk<void, void, AppThunkConfig>(
```
I think I added a comment for this, using a middleware to change the api provider when we dispatch a `setApiEnv` action will make it easier to handle also the problems that will create. Clear current state for the 
previous network etc.

pchrysochoidis commented on wallet/src/ui/styles/global.scss:
```diff
@@ -1,4 +1,5 @@
 @import url('bootstrap-icons');
+@import url('https://fonts.googleapis.com/css2?family=Space+Mono:wght@400;700&display=swap');
```
I think we could remove the custom font at the moment, since we expect to have new designs. If you want to use it please change it to include the font from a local resource

pchrysochoidis commented on wallet/src/ui/app/redux/slices/txresults/index.ts:
```diff
@@ -0,0 +1,154 @@
+} from '@mysten/sui.js';
+import type { AppThunkConfig } from '_store/thunk-extras';
+
+type TxResultState = {
```
We could export this and avoid re-defining it [here](https://github.com/MystenLabs/sui/pull/2529/files#diff-53b58dfabcda4999bb9f97abaf2cfa9f3d528c57631a0dd38c8bd7de06d6af28R14). It's the same right?

pchrysochoidis commented on wallet/src/ui/app/redux/slices/txresults/index.ts:
```diff
@@ -0,0 +1,154 @@
+});
+type TxResultByAddress = TxResultState[];
+
+const deduplicate = (results: [number, string][] | undefined) =>
```
I am not sure if this is necessary. Did you check if the adapter overwrites the txs with the same id already?

pchrysochoidis commented on wallet/src/ui/app/redux/slices/txresults/index.ts:
```diff
@@ -0,0 +1,154 @@
+                        // Remove failed transactions and sort by sequence number
+                        .filter((itm) => itm)
+                        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+                        .sort((a, b) => b!.seq - a!.seq)
```
The adapter has a [sorting functionality](https://redux-toolkit.js.org/api/createEntityAdapter#sortcomparer) as well . I guess it's the same for now that we only populate it with this action

pchrysochoidis commented on wallet/src/ui/app/redux/slices/app/index.ts:
```diff
@@ -1,24 +1,35 @@
 
 type AppState = {
     appType: AppType;
     apiEnv: API_ENV | null;
```
Removing this `| null` should help with types as well. It's never null anyway

pchrysochoidis commented on wallet/src/ui/app/redux/slices/txresults/index.ts:
```diff
@@ -0,0 +1,154 @@
+    results
+        ? results
+              .map((result) => result[1])
+              .filter((value, index, self) => self.indexOf(value) === index)
```
Might be worth it to add a comment why we need this. Why do we get the same transaction id multiple times?

pchrysochoidis commented on wallet/src/ui/app/components/transactions-card/index.tsx:
```diff
@@ -0,0 +1,96 @@
+
+function TransactionResult({ txresults }: { txresults: TxnData[] }) {
+    return (
+        <div className={st['tx-container']}>
```
After [this](https://github.com/MystenLabs/sui/commit/a7ec54bed3476de76c1d72779d343ea55261162e#diff-0d73a2b9d22dfdd04ea4dd097a5cdb125ed232bd9f164074bce9f2af787bd69bR115) all class names exported from modules will be in camel case as well eg. `txContainer`. I think it makes it a bit easier to use. (At some point we can refactor all usages and then export only the camelCase format)

Romainua commented:
Will be close test for wallet?

Jibz1 commented:
> Will be close test for wallet? 

Yes, mostly minor changes.

Jibz1 commented:
@pchrysochoidis I addressed most of your concerns stated above. Review 

https://user-images.githubusercontent.com/8676844/173573831-3d4a362d-c594-4d85-b926-aeb408037ee7.mov
